bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java

```diff
@@ -0,0 +1,203 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link TouchWandWebSockets} class implements WebSockets API to TouchWand controller
+ *
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault({})
+public class TouchWandWebSockets {
+
+    private WebSocketClient client;
+    private String controllerAddress;
+    private TouchWandSocket touchWandSocket;
+    private boolean isShutDown = false;
+
+    private final Logger logger = LoggerFactory.getLogger(TouchWandWebSockets.class);
+    private List<TouchWandUnitStatusUpdateListener> listeners = new ArrayList<>();
+    @Nullable
+    private ScheduledFuture<?> socketReconnect;
+
+    private static final String WS_ENDPOINT_TOUCHWAND = "/async";
+    private @Nullable URI uri;
+    private static final int CONNECT_TIMEOUT = 10000; // 10 seconds
+    private static final int WEBSOCKET_RECONNECT_INTERVAL = CONNECT_TIMEOUT * 2;
+    private static final int WEBSOCKET_IDLE_TIMEOUT = CONNECT_TIMEOUT * 10;
+    ScheduledExecutorService scheduler;
+
+    public TouchWandWebSockets(String ipAddress, ScheduledExecutorService scheduler) {
+        client = new WebSocketClient();
+        touchWandSocket = new TouchWandSocket();
+        this.controllerAddress = ipAddress;
+        this.scheduler = scheduler;
+        socketReconnect = null;
+    }
+
+    public void connect() {
+        try {
+            uri = new URI("ws://" + controllerAddress + WS_ENDPOINT_TOUCHWAND);
+        } catch (URISyntaxException e) {
+            logger.warn("URI not valid {} message {}", uri, e.getMessage());
+            return;
+        }
+
+        client.setConnectTimeout(CONNECT_TIMEOUT);
+        ClientUpgradeRequest request = new ClientUpgradeRequest();
+        request.setSubProtocols("relay_protocol");
+
+        try {
+            client.start();
+            client.connect(touchWandSocket, uri, request);
+        } catch (Exception e) {
+            logger.warn("Could not connect webSocket URI {} message {}", uri, e.getMessage());
+            return;
+        }
+    }
+
+    public void dispose() {
+        isShutDown = true;
+        try {
+            client.stop();
+        } catch (Exception e) {
+            logger.warn("Could not stop webSocketClient,  message {}", e.getMessage());
+        }
+    }
+
+    public synchronized void registerListener(TouchWandUnitStatusUpdateListener listener) {
+        if (!listeners.contains(listener)) {
+            logger.debug("Adding TouchWandWebSocket listener {}", listener);
+            listeners.add(listener);
+        }
+    }
+
+    public synchronized void unregisterListener(TouchWandUnitStatusUpdateListener listener) {
+        logger.debug("Removing TouchWandWebSocket listener {}", listener);
+        listeners.remove(listener);
+    }
+
+    @WebSocket(maxIdleTime = WEBSOCKET_IDLE_TIMEOUT)
+    @NonNullByDefault
```

Comment:
Inner classes don't need to be annotated with NonNull. There is a bug that checkstyle complains about it. You can safely ignore it.